src/script/tari_script.rs

```diff
@@ -462,53 +468,40 @@ impl TariScript {
         }
     }
 
-    fn check_multisig(&self, stack: &mut ExecutionStack, message: Message) -> Result<bool, ScriptError> {
-        use StackItem::*;
-        // get n
-        // must be gt 0
-        let n = match stack.pop().ok_or(ScriptError::StackUnderflow)? {
-            Number(n) if n > 0 => n as usize,
-            Number(_) => return Err(ScriptError::InvalidInput),
-            _ => return Err(ScriptError::IncompatibleTypes),
-        };
-        // pop n pubkeys
-        let public_keys = stack
-            .pop_num_items(n)?
-            .into_iter()
-            .map(|item| match item {
-                PublicKey(pk) => Ok(pk),
-                _ => Err(ScriptError::IncompatibleTypes),
-            })
-            .collect::<Result<Vec<RistrettoPublicKey>, ScriptError>>()?;
-
-        // get m
-        // must be gt 0 and lte n
-        let m = match stack.pop().ok_or(ScriptError::StackUnderflow)? {
-            Number(m) if m > 0 && m <= n as i64 => m as usize,
-            Number(_) => return Err(ScriptError::InvalidInput),
-            _ => return Err(ScriptError::IncompatibleTypes),
-        };
+    fn check_multisig(
+        &self,
+        stack: &mut ExecutionStack,
+        m: u8,
+        n: u8,
+        public_keys: &Vec<RistrettoPublicKey>,
+        message: Message,
+    ) -> Result<bool, ScriptError> {
+        if m == 0 || n == 0 || m > n || n > MAX_MULTISIG_LIMIT {
+            return Err(ScriptError::InvalidData);
+        }
         // pop m sigs
         let signatures = stack
-            .pop_num_items(m)?
+            .pop_num_items(m.into())?
             .into_iter()
             .map(|item| match item {
-                Signature(s) => Ok(s),
+                StackItem::Signature(s) => Ok(s),
                 _ => Err(ScriptError::IncompatibleTypes),
             })
             .collect::<Result<Vec<RistrettoSchnorr>, ScriptError>>()?;
 
-        let mut valid_sigs = 0;
-        for signature in signatures.iter() {
-            for public_key in public_keys.iter() {
-                if signature.verify_challenge(public_key, &message) {
-                    valid_sigs += 1;
+        let mut valid_signatures = 0;
+        let mut key_signed = vec![false; public_keys.len()];
+        for (i, public_key) in public_keys.iter().enumerate() {
```

Comment:
I suggest switching these two loops, then you can fail fast as soon an invalid signature is encountered.